NPE when msgstore.db is the input evidence

Three processing presented the following error:

**1.**
```text
2025-06-25 13:17:07     [ERROR] [parsers.whatsapp.WhatsAppParser]                       Error parsing WhatsApp: Item: msgstore.db type: sqlite size: 156393472
org.apache.tika.exception.TikaException: WAExtractorException Exception
        at iped.parsers.whatsapp.WhatsAppParser.parseDB(WhatsAppParser.java:424) ~[iped-parsers-impl-4.2.2.jar:?]
        at iped.parsers.whatsapp.WhatsAppParser.parseAndCheckIfIsMainDb(WhatsAppParser.java:479) ~[iped-parsers-impl-4.2.2.jar:?]
        at iped.parsers.whatsapp.WhatsAppParser.parse(WhatsAppParser.java:251) [iped-parsers-impl-4.2.2.jar:?]
        at org.apache.tika.parser.CompositeParser.parse(CompositeParser.java:298) [tika-core-2.4.0-p1.jar:2.4.0]
        at iped.parsers.standard.StandardParser.parse(StandardParser.java:245) [iped-parsers-impl-4.2.2.jar:?]
        at iped.engine.io.ParsingReader$BackgroundParsing.run(ParsingReader.java:247) [iped-engine-4.2.2.jar:?]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) [?:?]
        at java.util.concurrent.FutureTask.run(FutureTask.java:264) [?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
        at java.lang.Thread.run(Thread.java:834) [?:?]
Caused by: java.lang.NullPointerException
        at iped.parsers.whatsapp.WhatsAppParser.fillAccountAvatar(WhatsAppParser.java:838) ~[iped-parsers-impl-4.2.2.jar:?]
        at iped.parsers.whatsapp.WhatsAppParser.getUserAccount(WhatsAppParser.java:907) ~[iped-parsers-impl-4.2.2.jar:?]
        at iped.parsers.whatsapp.WhatsAppParser.parseDB(WhatsAppParser.java:414) ~[iped-parsers-impl-4.2.2.jar:?]
        ... 10 more
2025-06-25 13:17:10     [ERROR] [parsers.whatsapp.WhatsAppParser]                       Error parsing WhatsApp: Item: msgstore.db type: sqlite size: 156393472
org.apache.tika.exception.TikaException: WAExtractorException Exception
        at iped.parsers.whatsapp.WhatsAppParser.parseDB(WhatsAppParser.java:424) ~[iped-parsers-impl-4.2.2.jar:?]
        at iped.parsers.whatsapp.WhatsAppParser.parseAndCheckIfIsMainDb(WhatsAppParser.java:479) ~[iped-parsers-impl-4.2.2.jar:?]
        at iped.parsers.whatsapp.WhatsAppParser.parse(WhatsAppParser.java:251) [iped-parsers-impl-4.2.2.jar:?]
        at org.apache.tika.parser.CompositeParser.parse(CompositeParser.java:298) [tika-core-2.4.0-p1.jar:2.4.0]
        at iped.parsers.standard.StandardParser.parse(StandardParser.java:245) [iped-parsers-impl-4.2.2.jar:?]
        at iped.engine.task.MakePreviewTask$1.run(MakePreviewTask.java:181) [iped-engine-4.2.2.jar:?]
Caused by: java.lang.NullPointerException
        at iped.parsers.whatsapp.WhatsAppParser.fillAccountAvatar(WhatsAppParser.java:838) ~[iped-parsers-impl-4.2.2.jar:?]
        at iped.parsers.whatsapp.WhatsAppParser.getUserAccount(WhatsAppParser.java:907) ~[iped-parsers-impl-4.2.2.jar:?]
        at iped.parsers.whatsapp.WhatsAppParser.parseDB(WhatsAppParser.java:414) ~[iped-parsers-impl-4.2.2.jar:?]
        ... 5 more
```

**2.**
```text
2025-06-25 13:18:06     [ERROR] [parsers.whatsapp.WhatsAppParser]                       Error parsing WhatsApp: Item: msgstore.db type: sqlite size: 202391552
org.apache.tika.exception.TikaException: WAExtractorException Exception
        at iped.parsers.whatsapp.WhatsAppParser.parseDB(WhatsAppParser.java:424) ~[iped-parsers-impl-4.2.2.jar:?]
        at iped.parsers.whatsapp.WhatsAppParser.parseAndCheckIfIsMainDb(WhatsAppParser.java:479) ~[iped-parsers-impl-4.2.2.jar:?]
        at iped.parsers.whatsapp.WhatsAppParser.parse(WhatsAppParser.java:251) [iped-parsers-impl-4.2.2.jar:?]
        at org.apache.tika.parser.CompositeParser.parse(CompositeParser.java:298) [tika-core-2.4.0-p1.jar:2.4.0]
        at iped.parsers.standard.StandardParser.parse(StandardParser.java:245) [iped-parsers-impl-4.2.2.jar:?]
        at iped.engine.io.ParsingReader$BackgroundParsing.run(ParsingReader.java:247) [iped-engine-4.2.2.jar:?]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) [?:?]
        at java.util.concurrent.FutureTask.run(FutureTask.java:264) [?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
        at java.lang.Thread.run(Thread.java:834) [?:?]
Caused by: java.lang.NullPointerException
        at iped.parsers.whatsapp.WhatsAppParser.fillAccountAvatar(WhatsAppParser.java:838) ~[iped-parsers-impl-4.2.2.jar:?]
        at iped.parsers.whatsapp.WhatsAppParser.getUserAccount(WhatsAppParser.java:907) ~[iped-parsers-impl-4.2.2.jar:?]
        at iped.parsers.whatsapp.WhatsAppParser.parseDB(WhatsAppParser.java:414) ~[iped-parsers-impl-4.2.2.jar:?]
        ... 10 more
2025-06-25 13:18:10     [ERROR] [parsers.whatsapp.WhatsAppParser]                       Error parsing WhatsApp: Item: msgstore.db type: sqlite size: 202391552
org.apache.tika.exception.TikaException: WAExtractorException Exception
        at iped.parsers.whatsapp.WhatsAppParser.parseDB(WhatsAppParser.java:424) ~[iped-parsers-impl-4.2.2.jar:?]
        at iped.parsers.whatsapp.WhatsAppParser.parseAndCheckIfIsMainDb(WhatsAppParser.java:479) ~[iped-parsers-impl-4.2.2.jar:?]
        at iped.parsers.whatsapp.WhatsAppParser.parse(WhatsAppParser.java:251) [iped-parsers-impl-4.2.2.jar:?]
        at org.apache.tika.parser.CompositeParser.parse(CompositeParser.java:298) [tika-core-2.4.0-p1.jar:2.4.0]
        at iped.parsers.standard.StandardParser.parse(StandardParser.java:245) [iped-parsers-impl-4.2.2.jar:?]
        at iped.engine.task.MakePreviewTask$1.run(MakePreviewTask.java:181) [iped-engine-4.2.2.jar:?]
Caused by: java.lang.NullPointerException
        at iped.parsers.whatsapp.WhatsAppParser.fillAccountAvatar(WhatsAppParser.java:838) ~[iped-parsers-impl-4.2.2.jar:?]
        at iped.parsers.whatsapp.WhatsAppParser.getUserAccount(WhatsAppParser.java:907) ~[iped-parsers-impl-4.2.2.jar:?]
        at iped.parsers.whatsapp.WhatsAppParser.parseDB(WhatsAppParser.java:414) ~[iped-parsers-impl-4.2.2.jar:?]
        ... 5 more
```

**3.**
```text
2025-06-25 13:19:08     [ERROR] [parsers.whatsapp.WhatsAppParser]                       Error parsing WhatsApp: Item: msgstore.db type: sqlite size: 80568320
org.apache.tika.exception.TikaException: WAExtractorException Exception
        at iped.parsers.whatsapp.WhatsAppParser.parseDB(WhatsAppParser.java:424) ~[iped-parsers-impl-4.2.2.jar:?]
        at iped.parsers.whatsapp.WhatsAppParser.parseAndCheckIfIsMainDb(WhatsAppParser.java:479) ~[iped-parsers-impl-4.2.2.jar:?]
        at iped.parsers.whatsapp.WhatsAppParser.parse(WhatsAppParser.java:251) [iped-parsers-impl-4.2.2.jar:?]
        at org.apache.tika.parser.CompositeParser.parse(CompositeParser.java:298) [tika-core-2.4.0-p1.jar:2.4.0]
        at iped.parsers.standard.StandardParser.parse(StandardParser.java:245) [iped-parsers-impl-4.2.2.jar:?]
        at iped.engine.io.ParsingReader$BackgroundParsing.run(ParsingReader.java:247) [iped-engine-4.2.2.jar:?]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) [?:?]
        at java.util.concurrent.FutureTask.run(FutureTask.java:264) [?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
        at java.lang.Thread.run(Thread.java:834) [?:?]
Caused by: java.lang.NullPointerException
        at iped.parsers.whatsapp.WhatsAppParser.fillAccountAvatar(WhatsAppParser.java:838) ~[iped-parsers-impl-4.2.2.jar:?]
        at iped.parsers.whatsapp.WhatsAppParser.getUserAccount(WhatsAppParser.java:907) ~[iped-parsers-impl-4.2.2.jar:?]
        at iped.parsers.whatsapp.WhatsAppParser.parseDB(WhatsAppParser.java:414) ~[iped-parsers-impl-4.2.2.jar:?]
        ... 10 more
2025-06-25 13:19:10     [ERROR] [parsers.whatsapp.WhatsAppParser]                       Error parsing WhatsApp: Item: msgstore.db type: sqlite size: 80568320
org.apache.tika.exception.TikaException: WAExtractorException Exception
        at iped.parsers.whatsapp.WhatsAppParser.parseDB(WhatsAppParser.java:424) ~[iped-parsers-impl-4.2.2.jar:?]
        at iped.parsers.whatsapp.WhatsAppParser.parseAndCheckIfIsMainDb(WhatsAppParser.java:479) ~[iped-parsers-impl-4.2.2.jar:?]
        at iped.parsers.whatsapp.WhatsAppParser.parse(WhatsAppParser.java:251) [iped-parsers-impl-4.2.2.jar:?]
        at org.apache.tika.parser.CompositeParser.parse(CompositeParser.java:298) [tika-core-2.4.0-p1.jar:2.4.0]
        at iped.parsers.standard.StandardParser.parse(StandardParser.java:245) [iped-parsers-impl-4.2.2.jar:?]
        at iped.engine.task.MakePreviewTask$1.run(MakePreviewTask.java:181) [iped-engine-4.2.2.jar:?]
Caused by: java.lang.NullPointerException
        at iped.parsers.whatsapp.WhatsAppParser.fillAccountAvatar(WhatsAppParser.java:838) ~[iped-parsers-impl-4.2.2.jar:?]
        at iped.parsers.whatsapp.WhatsAppParser.getUserAccount(WhatsAppParser.java:907) ~[iped-parsers-impl-4.2.2.jar:?]
        at iped.parsers.whatsapp.WhatsAppParser.parseDB(WhatsAppParser.java:414) ~[iped-parsers-impl-4.2.2.jar:?]
        ... 5 more
```

**EDIT**: Just changed the format of error messages to make them more readable. (@wladimirleite)

I only processed the directories with the msgstore.db files.

Indeed, the 1st case, the error is that the file is corrupted and connot be opened in the SQL editor.

The 2nd case, successfuly two times processed, no errors. \o/

The 3rd case:

```
java -jar c:\IPED\IPED-4.2.2_and_java_plugins\iped-4.2.2\iped.jar -d D:\JP3\59_25\Eq03 -o D:\JP3\59_25-Eq03-IPED --portable
2025-06-27 14:20:21     [ERROR] [parsers.whatsapp.WhatsAppParser]                       Error parsing WhatsApp: Item: Eq03/msgstore[00001].db type: sqlite size: 41533440
org.apache.tika.exception.TikaException: WAExtractorException Exception
        at iped.parsers.whatsapp.WhatsAppParser.mergeParsedDBsAndOutputResults(WhatsAppParser.java:746) ~[iped-parsers-impl-4.2.2.jar:?]
        at iped.parsers.whatsapp.WhatsAppParser.parse(WhatsAppParser.java:265) [iped-parsers-impl-4.2.2.jar:?]
        at org.apache.tika.parser.CompositeParser.parse(CompositeParser.java:298) [tika-core-2.4.0-p1.jar:2.4.0]
        at iped.parsers.standard.StandardParser.parse(StandardParser.java:245) [iped-parsers-impl-4.2.2.jar:?]
        at iped.engine.io.ParsingReader$BackgroundParsing.run(ParsingReader.java:247) [iped-engine-4.2.2.jar:?]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) [?:?]
        at java.util.concurrent.FutureTask.run(FutureTask.java:264) [?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
        at java.lang.Thread.run(Thread.java:834) [?:?]
Caused by: java.util.ConcurrentModificationException
        at java.util.ArrayList.sort(ArrayList.java:1752) ~[?:?]
        at java.util.Collections.sort(Collections.java:145) ~[?:?]
        at iped.parsers.whatsapp.ReportGenerator.printMessage(ReportGenerator.java:1067) ~[iped-parsers-impl-4.2.2.jar:?]
        at iped.parsers.whatsapp.ReportGenerator.lambda$generateNextChatHtml$0(ReportGenerator.java:230) ~[iped-parsers-impl-4.2.2.jar:?]
        at iped.parsers.whatsapp.ReportGenerator$1.lookup(ReportGenerator.java:1665) ~[iped-parsers-impl-4.2.2.jar:?]
        at org.apache.commons.text.StringSubstitutor.resolveVariable(StringSubstitutor.java:1148) ~[commons-text-1.10.0.jar:1.10.0]
        at org.apache.commons.text.StringSubstitutor.substitute(StringSubstitutor.java:1514) ~[commons-text-1.10.0.jar:1.10.0]
        at org.apache.commons.text.StringSubstitutor.substitute(StringSubstitutor.java:1389) ~[commons-text-1.10.0.jar:1.10.0]
        at org.apache.commons.text.StringSubstitutor.replace(StringSubstitutor.java:893) ~[commons-text-1.10.0.jar:1.10.0]
        at iped.parsers.whatsapp.ReportGenerator.printMessageFile(ReportGenerator.java:1678) ~[iped-parsers-impl-4.2.2.jar:?]
        at iped.parsers.whatsapp.ReportGenerator.generateNextChatHtml(ReportGenerator.java:202) ~[iped-parsers-impl-4.2.2.jar:?]
        at iped.parsers.whatsapp.WhatsAppParser.createReport(WhatsAppParser.java:299) ~[iped-parsers-impl-4.2.2.jar:?]
        at iped.parsers.whatsapp.WhatsAppParser.mergeParsedDBsAndOutputResults(WhatsAppParser.java:737) ~[iped-parsers-impl-4.2.2.jar:?]
        ... 9 more

IPED finished.
Check the log at C:\IPED\IPED-4.2.2_and_java_plugins\iped-4.2.2\log\IPED-2025-06-27-14-20-07.log
```

--- a/iped-parsers/iped-parsers-impl/src/main/java/iped/parsers/whatsapp/WhatsAppParser.java
+++ b/iped-parsers/iped-parsers-impl/src/main/java/iped/parsers/whatsapp/WhatsAppParser.java
@@ -872,7 +872,7 @@ public class WhatsAppParser extends SQLite3DBParser {
     }
 
     private void fillAccountAvatar(WAAccount account, IItemSearcher searcher, String dbPath) {
-        if (account == null || account.getAvatar() != null || account.getAvatarPath() != null || dbPath == null) {
+        if (searcher == null || account == null || account.getAvatar() != null || account.getAvatarPath() != null || dbPath == null) {
             return;
         }
 

